pysmt/optimization/optimizer.py

```diff
@@ -151,34 +202,190 @@ def optimize(self, cost_function, strategy='ub',
 
         `step_size` the minimum reolution for finding a solution. The
         optimum will be found in the proximity of `step_size`
-
         """
+        rt = None, None
+        if goal.is_simple_goal():
+            rt = self._simple_optimize(goal, strategy,
+                             feasible_solution_callback,
+                             step_size, **kwargs)
+        else:
+            raise GoalNotSupportedError("ExternalOptimizerMixin", goal)
+        return rt
+
+
+
+    def _simple_optimize(self, goal, strategy,
+                      feasible_solution_callback,
+                      step_size, **kwargs):
+        rt = None
+        if goal.is_maximization_goal():
+            rt = self._optimize_max(goal, strategy,
+                             feasible_solution_callback,
+                             step_size, **kwargs)
+        elif goal.is_minimization_goal():
+            rt = self._optimize_min(goal, strategy,
+                             feasible_solution_callback,
+                             step_size, **kwargs)
+        elif goal.is_minmax_goal():
+            rt = self._optimize_minmax(goal, strategy,
+                                    feasible_solution_callback,
+                                    step_size, **kwargs)
+        elif goal.is_maxmin_goal():
+            rt = self._optimize_maxmin(goal, strategy,
+                                    feasible_solution_callback,
+                                    step_size, **kwargs)
+        else:
+            raise GoalNotSupportedError("ExternalOptimizerMixin", goal)
+        return rt
+
+    def _optimize_minmax(self, goal, strategy,
+                 feasible_solution_callback,
+                 step_size, **kwargs):
+
+        part_res = []
+
+        for term in goal.terms():
+            self.push()
+            part_res.append(self._optimize_min(MinimizationGoal(term),
```

Comment:
Nope. This is not the intended semantics!
have a look here (http://disi.unitn.it/rseba/papers/cav15_extended.pdf, page 4):
![Screenshot from 2020-05-04 10-25-47](https://user-images.githubusercontent.com/2255972/80948095-b74af880-8df1-11ea-8d1f-6b5ef447a35f.png)

Reply:
Note also that pysmt has a `Max` and a `Min` operator that symbolically express the maximum and the minimum value among a gven, finite set. This could be useful for a simple initial implementation...

Reply:
> Nope. This is not the intended semantics!
> have a look here (http://disi.unitn.it/rseba/papers/cav15_extended.pdf, page 4):
> ![Screenshot from 2020-05-04 10-25-47](https://user-images.githubusercontent.com/2255972/80948095-b74af880-8df1-11ea-8d1f-6b5ef447a35f.png)

Just a comment to repeat what I stated in the email: the above definition is for Max-Min, the one for Min-Max is dual.

Reply:
> Note also that pysmt has a `Max` and a `Min` operator that symbolically express the maximum and the minimum value among a gven, finite set. This could be useful for a simple initial implementation...

The above definition is more convenient than using a `max`/`min` operator, because the objective function is always *explicitly* upper-bounded (or lower-bounded, depending on the definition) by the terms in the given list of arguments.